### ISSUE TITLE
Tracking Issue for result_ffi_guarantees (RFC 3391)

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#110503]
This is a tracking issue for the RFC "result_ffi_guarantees" (rust-lang/rfcs#3391).

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement (test) the RFC
  - [ ] https://github.com/rust-lang/rust/pull/122253
  - [x] https://github.com/rust-lang/rust/pull/124747
  - [x] https://github.com/rust-lang/rust/pull/124870
  - [x] https://github.com/rust-lang/rust/pull/125672
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [x] ~Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])~ (not necessary, reference FCP should be all that's needed)

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

None

**Comment by RalfJung** [comment]
I did a bunch of work on ABI tests for `repr(transparent)` that could also be useful for testing this RFC:
- Miri now has this test: https://github.com/rust-lang/miri/blob/master/tests/pass/function_calls/abi_compat.rs
- and rustct will hopefully soon have this: https://github.com/rust-lang/rust/pull/115372

**Comment by MasterAwesome** [comment]
Is anyone still working on this? If not, I'd love to try my shot at it.

**Comment by Lokathor** [comment]
I have not heard word of anyone working on this in a very long time. It's probably a safe bet to begin a PR.

**Comment by RalfJung** [comment]
Besides adjusting the improper-ctypes lint (which is happening in https://github.com/rust-lang/rust/pull/122253), some other things that should be done:
- For `Option` we have [this](https://doc.rust-lang.org/nightly/std/option/index.html#representation). I assume we want the same for `Result`? Is there some way to not repeat the table?
- Miri has a test [here](https://github.com/rust-lang/miri/blob/e28045d5811a4eb51cf3b3d66c055b4ab02c1ff2/tests/pass/function_calls/abi_compat.rs#L86-L91) that currently only checks `Option`, not `Result`.

**Comment by MasterAwesome** [comment]
Should the documentation change be a separate PR?

**Comment by RalfJung** [comment]
Yes, that probably makes review easier.

**Comment by Lokathor** [comment]
The lint update PR has merged.

**Comment by RalfJung** [comment]
I hope docs are next, then. :) We shouldn't go too long with such a gap between what we document and what we implement.

The `Result` type itself should get suitable documentation. And probably there's somewhere in the reference as well that needs updating?

**Comment by Lokathor** [comment]
philosophically speaking I'm not sure how much the reference needs to say in addition to what the standard library type docs say, but yeah docs on Option and Result are the next step. I will likely have time for a docs PR very soon if no one else does it first

**Comment by Lokathor** [comment]
The docs are merged.

@tmandry Is there more to do or can we officially call this done?

**Comment by RalfJung** [comment]
I think Miri still needs tests. Those would go around [here](https://github.com/rust-lang/rust/blob/bd8e565e16b43e2e68a67804d20992218579ac9d/src/tools/miri/tests/pass/function_calls/abi_compat.rs#L86-L91).

**Comment by Lokathor** [comment]
Followup: The MIRI test cases have been merged.

I believe we're done? Should we put this into the release notes two releases from now?

**Comment by RalfJung** [comment]
Yeah sounds right!
I have added the relnotes label to https://github.com/rust-lang/rust/pull/124870. But with that PR being "closed" rather than "merged" I don't know if that will work properly...

**Comment by cuviper** [comment]
This is still unstable, right? What exactly are we wanting to promise in relnotes?

e.g. This still warns if the feature gate is not enabled: [playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=928efa649a01ae416e3141d8643c7ae9)

```rust
//#![feature(result_ffi_guarantees)]

extern "C" {
    pub fn foo() -> Result<(), std::num::NonZeroU32>;
}
```

```
warning: `extern` block uses type `Result<(), NonZero<u32>>`, which is not FFI-safe
 --> src/lib.rs:4:21
  |
4 |     pub fn foo() -> Result<(), std::num::NonZeroU32>;
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
  |
  = help: consider adding a `#[repr(C)]`, `#[repr(transparent)]`, or integer `#[repr(...)]` attribute to this enum
  = note: enum has no representation hint
  = note: `#[warn(improper_ctypes)]` on by default
```

**Comment by Lokathor** [comment]
I guess there needs to be a final stabilization PR for the feature, but it doesn't need T-lang signoff?

**Comment by RalfJung** [comment]
Why would it not need t-lang signoff?

**Comment by Lokathor** [comment]
that's just what @tmandry wrote in the start of the thread next to the checked box.

**Comment by RalfJung** [comment]
Where is the reference PR they are mentioning there?

**Comment by Lokathor** [comment]
I don't know, actually. I just saw a checked box and assumed the box wouldn't be checked if it wasn't completed yet.

**Comment by tmandry** [comment]
I must have been under the impression that this was only a documentation change. It probably needs a lang team signoff since it's more than that.

**Comment by workingjubilee** [comment]
Hm? Lang FCPed it? https://github.com/rust-lang/rfcs/pull/3391#issuecomment-1448763059

**Comment by RalfJung** [comment]
Indeed the change already [landed](https://github.com/rust-lang/rust/pull/124747) and hit stable with Rust 1.80.

The only thing that's missing, AFAIK, is updating the Reference? Or maybe that has been done already but we forgot to close this issue?
Cc @ehuss

**Comment by cuviper** [comment]
The feature gate is still unstable:

https://github.com/rust-lang/rust/blob/1a5a2240bc1b8cf0bcce7acb946c78d6493a4fd3/compiler/rustc_feature/src/unstable.rs#L575-L577

That needs to move to accepted in the neighboring module, and might need updates in places that use it too, like:

https://github.com/rust-lang/rust/blob/1a5a2240bc1b8cf0bcce7acb946c78d6493a4fd3/compiler/rustc_lint/src/types.rs#L1147-L1149

**Comment by workingjubilee** [comment]
I have opened #130628 on the assumption that all necessary documentation changes have landed already. Please let me know if that is incorrect.

**Comment by ehuss** [comment]
I do not believe this has been documented. I think that will be needed before stabilization should continue.

Also, it would be helpful to update the top comment with links to the implementation, since it's not really clear to me what was actually changed.

**Comment by workingjubilee** [comment]
I believe it was documented on the Result type. Should we add it to the Reference PR?

**Comment by workingjubilee** [comment]
Indeed the Result type was documented accordingly: https://doc.rust-lang.org/std/result/index.html#representation

**Comment by workingjubilee** [comment]
I have lifted all PRs relevant to the implementation history to the top.

**Comment by workingjubilee** [comment]
@ehuss It seems none of the ABI guarantees of Result or Option have been documented in the Reference that I can see, and I'm not sure where the other ABI guarantees that we offer are in the Reference either. So I think this is a larger concern than the matter of this feature and a separate issue should be opened for it. [Type layout](https://doc.rust-lang.org/stable/reference/type-layout.html) doesn't seem even close to exhaustive.

**Comment by workingjubilee** [comment]
@ehuss I *can* post a bare-minimum PR to the reference but it will mostly serve to make it more obvious that the work is direly unfinished here.

**Comment by ehuss** [comment]
Thanks @workingjubilee! 

I am a bit confused, since I don't fully understand the connection between `result_ffi_guarantees` versus the already documented behavior in https://doc.rust-lang.org/std/result/index.html#representation. Is that current documentation making a stronger statement than what is actually true today? Or is there something missing?

I am also quite confused about what is being stabilized, or about the process being used. From the timeline I see:
1. https://github.com/rust-lang/rfcs/pull/3391 was merged to guarantee certain layouts for `Result` and `Option` used in FFI.
2. https://github.com/rust-lang/rust/pull/124747 implemented this for `Result` for FFI under the `result_ffi_guarantees` feature flag.
3. https://github.com/rust-lang/rust/pull/124870 added documentation for `Result`, even though it is still behind a feature flag?

Some questions:
* Does `result_ffi_guarantees` actually change the layout, or is it just related to squelching the `improper-ctypes` lint?
* Why was the layout guarantee documented before it was stabilized? Is there some difference between repr-Rust (guaranteed on stable) and repr-C (not niche-optimized on stable)?
* Where is the stabilization report, and the t-lang approval of that?
* The RFC mentions `Option`, but the two PRs above seem to only be for `Result`. What happened with `Option`?

Sorry for all the dumb questions, but I am very confused as to what is being proposed.

As for layout guarantees, since the reference currently doesn't cover that, then I think it is fine to skip it for now. We are inching our way forward with https://github.com/rust-lang/reference/pull/1545, and I think this might be included there (cc @chorman0773).

**Comment by chorman0773** [comment]
> As for layout guarantees, since the reference currently doesn't cover that, then I think it is fine to skip it for now. We are inching our way forward with https://github.com/rust-lang/reference/pull/1545, and I think this might be included there (cc @chorman0773).

Will probably also want to go into the type-layout chapter as well. It can be added to abi as well once 1545 is merged.

**Comment by workingjubilee** [comment]
> I am a bit confused, since I don't fully understand the connection between result_ffi_guarantees versus the already documented behavior in https://doc.rust-lang.org/std/result/index.html#representation. Is that current documentation making a stronger statement than what is actually true today? Or is there something missing?

That IS the behavior with this feature.

> The RFC mentions Option, but the two PRs above seem to only be for Result. What happened with Option?

Option already had such an ABI guarantee documented: https://doc.rust-lang.org/std/option/index.html#representation

> Does `result_ffi_guarantees` actually change the layout, or is it just related to squelching the improper-ctypes lint?

My understanding is that it is not a functional behavioral change. It **is** a promise that the behavior will not change, and that if it varies from the documented behavior, it will be treated as a bug. To many developers, this is **much** more important than "how the compiler actually behaves".

> Why was the layout guarantee documented before it was stabilized? Is there some difference between repr-Rust (guaranteed on stable) and repr-C (not niche-optimized on stable)?

I don't understand the question. Result is not a `repr(C)` type so this is irrelevant?

**Comment by workingjubilee** [comment]
the RFC is not about:
- Option, which already is in the desired state.
- Other enums, which have no such guarantee.

the RFC is about:
- `Result<T, ()>` and `Result<(), T>` where the `T` formed allows a niche optimization to track the `Ok(())` or `Err(())` tag, and other ZST types in lieu of `()`.

**Comment by chorman0773** [comment]
FTR, I already have plans to redo the type-layout chapter, particularily the enum section. `Option` and `Result` will get their own few clauses in there.

**Comment by chorman0773** [comment]
Doing a stabilization report is probably a good idea nonetheless.

**Comment by workingjubilee** [comment]
I suppose I will draft one, then! No promises it won't be very silly though.

**Comment by ehuss** [comment]
> I don't understand the question. Result is not a `repr(C)` type so this is irrelevant?

Oh, sorry, yea, that didn't make sense. My thinking was, do we have a stable guarantee such that `let x: Result<NonZeroI32, ()> = unsafe { std::mem::transmute(12) };` is well-defined (all "Rust"-land), but `extern "C" { fn f() -> Result<NonZeroI32, ()>; }` is not (in "C" ABI land).

> gonna level with you here, this RFC was basically a stabilization report for a behavior we already implemented. 🙃

That's what I'm trying to drill down to; is this just a documentation change, or is it a behavior change? If it was a documentation change, then I would not have expected #122253 to exist, unless the only goal there was to squelch `improper-ctypes`. The description of the PR doesn't really clarify what it does, and the RFC never mentions that lint.

For documentation-only changes, we have generally *not* required a stabilization report or separate FCP for an accepted RFC. Documentation-only RFCs normally get their doc changes merged almost right away, and there is no tracking issue. This seems slightly more than that, though.

> I have no magical power to wave a wand and make them do that if they are not.

100% agree with you[^delegate], and I don't want to put up roadblocks where it's not appropriate. However, I do want to make sure that there is awareness and understanding of what is happening.

[^delegate]: You might be interested in https://github.com/rust-lang/lang-team/pull/290, which is a proposal to help scale some decision making.

**Comment by Lokathor** [comment]
@ehuss I don't believe we have any guarantee that something is *not* defined.

In fact the entire point of the RFC is to make your example cose snippit be defined, as far as i can read it

**Comment by workingjubilee** [comment]
@ehuss Of course! tbh I'm mostly just playing my occasional role of "clown that pratfalls into thing everyone is confused about, prompting everyone to cheer, boo, throw tomatoes, or buy popcorn". :^)

**Comment by workingjubilee** [comment]
> For documentation-only changes, we have generally not required a stabilization report or separate FCP for an accepted RFC. Documentation-only RFCs normally get their doc changes merged almost right away, and there is no tracking issue. This seems slightly more than that, though.

...speaking of documentation, is this written anywhere? I mean oral tradition is well and dandy but this actually seems worth formalizing because of the whole "T-lang has a tiny amount of bandwidth" 🫠 so it's good to note down when something only hard-requires _one_ Formal Team Decision instead of two Formal Team Decisions.

**Comment by workingjubilee** [comment]
I _do_ think a tracking issue sometimes is warranted simply because "wait how many places do we have to update the documentation exactly...?" sometimes uhh needs a Formal Issue.

And there _technically_ were related compiler changes.

**Comment by workingjubilee** [comment]
And yeah, with this feature, these are now both correct:
```rust
let x: Result<NonZeroI32, ()> = unsafe { std::mem::transmute(12) };
extern "C" {
    fn f() -> Result<NonZeroI32, ()>;
}
```

The latter even should, IIRC, be correct under the "integer normalization" CFI flag (which reduces entropy but enables certain ABI-compatible equivalencies). This is because the equivalence is sorta defined in terms of [transmute](https://doc.rust-lang.org/stable/std/primitive.fn.html#abi-compatibility):

> If the declared signature and the signature of the function pointer are ABI-compatible, then the function call behaves as if every argument was [transmuted](https://doc.rust-lang.org/stable/std/mem/fn.transmute.html) from the type in the function pointer to the type at the function declaration, and the return value is [transmuted](https://doc.rust-lang.org/stable/std/mem/fn.transmute.html) from the type in the declaration to the type in the pointer. All the usual caveats and concerns around transmutation apply; for instance, if the function expects a `NonZero<i32>` and the function pointer uses the ABI-compatible type `Option<NonZero<i32>>`, and the value used for the argument is None, then this call is Undefined Behavior since transmuting `None::<NonZero<i32>>` to `NonZero<i32>` violates the non-zero requirement.

**Comment by workingjubilee** [comment]
I have opened https://github.com/rust-lang/reference/issues/1623 to track "documenting a complete explanation of our ABI rules, also what even is an ABI, huh? what?"

**Comment by RalfJung** [comment]
The entire RFC was basically "just" a docs change. I didn't know there is a feature gate, all it can be doing is affecting the lint as feature gates affecting layout would be unsound so I hope we don't do it. ;)

So if this is anyway not in the reference, no update is needed there. Which means I think all that is left is stabilizing the feature - after we confirm what exactly it does.^^
